modulus/metrics/cae/integral.py

```diff
@@ -0,0 +1,130 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 - 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-FileCopyrightText: All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+import pyvista as pv
+
+
+def line_integral(
+    edges: np.ndarray,
+    points: np.ndarray,
+    field: np.ndarray,
+) -> float:
+    """
+    Compute integral along a line / edge. Currently assumes the curve in xy plane.
+
+    Parameters:
+    -----------
+    edges : np.ndarray
+        Edges of the curve in [M, 2] format
+    points : np.ndarray
+        Coordinates of points for the edge
+    field : np.ndarray
+        Field values at each edge center.
+
+    Returns:
+    --------
+    float
+        Integral along the given curve / line.
+    """
+    tangents = []
+    lengths = []
+
+    for i in range(edges.shape[0]):
+        vec = points[edges[i, 1]] - points[edges[i, 0]]
+        tangent = vec / np.linalg.norm(vec)
+        tangents.append(tangent)
+        lengths.append(np.linalg.norm(vec))
+
+    tangents = np.array(tangents)
+    lengths = np.array(lengths)
+
+    # integrate the results
+    if len(field.shape) == 2:
+        # Vector quantity
+        integral = np.sum(np.sum(field * tangents, axis=1) * lengths)
+    else:
+        # Scalar quantity
+        integral = np.sum(field * lengths)
+
+    return integral
+
+
+def surface_integral(
+    mesh: pv.PolyData,
```

Comment:
Shouldn't we also support `pv.UnstructuredGrid`?

Reply:
Good point, but probably we don't want to do that for this particular utility whose scope is to only compute surface integrals.

From pyvista's docs [here](https://docs.pyvista.org/api/core/), they mention:
> We generally use [pyvista.PolyData](https://docs.pyvista.org/api/core/_autosummary/pyvista.polydata#pyvista.PolyData) to construct scattered points and closed/open surfaces (non-volumetric datasets). 

Since the `pv.UnstructuredGrid` is a more general dataset format that allows 3d grids, we can potentially write a separate utility extracts such surfaces from the 3d mesh and then calls this utility for the integrals.